### PR TITLE
Update user-defined hash documentation

### PR DESCRIPTION
**Pull request by bradcray**
This is a follow-on to #19243 to document when the compiler
will or will not provide a hash function for a user record.  While
here, I also attempted to improve the existing documentation by:
* making the spec example for a user-defined hash testable and
  complete, fixing some typos
* providing a bit more context in the primer and removing the note
  about built-in types (since the primer is about special methods
  on classes and records and doesn't need to be as complete as
  the spec)
* word-smithing
* adding return types to the documentation examples of hash()

While here, I found myself questioning whether "returning `int`
or `uint`" was what we wanted to commit to, combined with
slight discomfort with the fact that the high bit is ignored in our
implementation without us saying so, so filed
https://github.com/chapel-lang/chapel/issues/19285 to capture
this concern.

### CODE REVIEW

File: doc/rst/language/spec/records.rst [hunk 1]
```diff
@@ -643,30 +643,46 @@ order they are declared in the record definition.
...
-   m[myR] = 5;
-
-   var myD = domain(R);
-   myD += myR;
-
-Note that the compiler generated ``hash`` can only be overriden on records
-that have been defined in user code, and will not override the compiler hash
-for ``int.hash``, for example.
+Note that the compiler-generated ``hash`` can only be overridden for
+records that have been defined in user code, and cannot be used to
```
Comment (mppf) [inline doc/rst/language/spec/records.rst:683]: For grammar - I recommend adding "it" after "and" or else removing the comma. Starting a new sentence would work too.

**Review by mppf — APPROVED** [review]

**Review by bmcdonald3 — APPROVED** [review]

File: doc/rst/language/spec/records.rst (hunk 1 above)
Comment (bradcray) [reply doc/rst/language/spec/records.rst]: Thanks, I split into two sentences.

File: doc/rst/language/spec/records.rst [hunk 2]
```diff
@@ -643,30 +643,46 @@ order they are declared in the record definition.
...
-.. code-block:: chapel
+   *Example (userhash.chpl)*.
+
+   .. code-block:: chapel
+
+
+      record R {
+        var i: uint;
+
+        proc hash() {
```
[the remark this answers is not in the document]
Comment (bradcray) [reply doc/rst/language/spec/records.rst]: Done.

File: test/release/examples/primers/specialMethods.chpl [hunk 3]
```diff
@@ -114,17 +114,24 @@ writeln(r.vals);
...
+// By default, the compiler will define a hash method for any record
+// that does not define its own ``==`` or ``!=`` overloads.  This
+// permits such records to be used as the indices of associative
+// domains, the values in a :mod:`Set`, or the keys in a :mod:`Map`.
+// Users can override this default by supplying their own hash method
+// that returns a ``uint`` or ``int`` value.  For example:
+
 use Map;
 
 proc R.hash() {
```
[the remark this answers is not in the document]
Comment (bradcray) [reply test/release/examples/primers/specialMethods.chpl]: Done.  I ended up filing #19285 as a result of this and the general uncertainty I ended up feeling about what return types/values we wanted to permit or not.

**Pull request merged**